posting/index.go

```diff
@@ -1387,39 +1386,86 @@ func rebuildTokIndex(ctx context.Context, rb *IndexRebuild) error {
 	builder.fn = func(uid uint64, pl *List, txn *Txn) ([]*pb.DirectedEdge, error) {
 		edge := pb.DirectedEdge{Attr: rb.Attr, Entity: uid}
 		edges := []*pb.DirectedEdge{}
-		err := pl.Iterate(txn.StartTs, 0, func(p *pb.Posting) error {
-			// Add index entries based on p.
-			val := types.Val{
-				Value: p.Value,
-				Tid:   types.TypeID(p.ValType),
-			}
-			edge.Lang = string(p.LangTag)
 
+		processAddIndexMutation := func(edge *pb.DirectedEdge, val types.Val) ([]*pb.DirectedEdge, error) {
 			for {
 				newEdges, err := txn.addIndexMutations(ctx, &indexMutationInfo{
 					tokenizers:   tokenizers,
 					factorySpecs: factorySpecs,
-					edge:         &edge,
+					edge:         edge,
 					val:          val,
 					op:           pb.DirectedEdge_SET,
 				})
 				switch err {
 				case ErrRetry:
 					time.Sleep(10 * time.Millisecond)
 				default:
-					if !runForVectors {
-						edges = append(edges, newEdges...)
-					}
-					return err
+					return newEdges, err
 				}
 			}
+		}
+
+		if runForVectors {
+			val, err := pl.Value(txn.StartTs)
+			if err != nil {
+				return []*pb.DirectedEdge{}, err
+			}
+
+			if val.Tid != types.VFloatID {
+				sv, err := types.Convert(val, types.VFloatID)
+				if err != nil {
+					return []*pb.DirectedEdge{}, err
+				}
+				b := types.ValueForType(types.BinaryID)
+				if err = types.Marshal(sv, &b); err != nil {
+					return []*pb.DirectedEdge{}, err
+				}
+				edge.Value = b.Value.([]byte)
+				edge.ValueType = types.VFloatID.Enum()
+
+				inKey := x.DataKey(edge.Attr, uid)
+				p, err := txn.Get(inKey)
+				if err != nil {
+					return []*pb.DirectedEdge{}, err
+				}
+
+				if err := p.addMutation(ctx, txn, &edge); err != nil {
+					return []*pb.DirectedEdge{}, err
+				}
+
+				_, err = processAddIndexMutation(&edge, val)
```

Comment:
this should be after the if condition. it should get executed for both when val.td == types.vfloatid.

Reply:
write a test that fails because of this bug right now, then fix the bug

Reply:
TestVectorIndexDropPredicate and TestVectorIndexWithoutSchema have already been added, and they fail on the main branch